File: scala/ch02/ex2.35.scala
def accumulate[A, B] (op: (A, B) => B, 
                      initial: B, sequence: List[A]): B =
  sequence match {
    case Nil => initial
    case _ => op (sequence.head, accumulate(op, initial, sequence.tail))
  }

def countLeaves (tree: List[_]): Int =
  accumulate ((node:Any, count:Int) => node match {
    case l:List[_] => count + countLeaves(l)
    case i:Int     => count + 1
    case _         => throw new RuntimeException("Invalid node type: "+node) 
  }, 0, tree)

val x  = List (List(1, 2), List(3, 4))
val xx = List (x, x)


import org.scalatest._ 
import org.scalatest.matchers._

object countLeavesSpec extends Spec with ShouldMatchers {
  describe ("countLeaves defined using accumulate") {
    it ("should return the number of leaf nodes in a tree") {
      countLeaves(x)  should equal (4)
      countLeaves(xx) should equal (8)
    }
  }
}
countLeavesSpec execute

File: clojure/ch02/ex2.35.clj
(ns sicp.ch02 (:use clojure.contrib.test-is))

(defn accumulate [op initial sequence]
  (if (empty? sequence)
      initial
      (op (first sequence)
          (accumulate op initial (rest sequence)))))

(defn count-leaves [tree]
  (accumulate (fn [node count]
                (if (not (list? node)) 
                    (+ count 1)
                    (+ count (count-leaves node)))) 
                0 tree))

(def x  (list (list 1 2) (list 3 4)))
(def xx (list x x))

(deftest test-length
  (is (= (count-leaves x)  4))
  (is (= (count-leaves xx) 8)))

(run-tests)
